### PR TITLE
DOC: Add expanded index descriptors for specifying for RangeIndex-as-metadata in Parquet file schema 

### LINKED ISSUE #25672: Expand Parquet pandas schema metadata to store RangeIndex without serialization
In https://github.com/pandas-dev/pandas/blob/master/doc/source/development/developer.rst, there is no affordance for storing RangeIndex without serializing it to a column of integers. This wastes both memory and time

I'll propose an evolution of the metadata that permits "non-serialized" indexes like RangeIndex to be stored without a conversion step of some kind

This will have to mind _forward_ compatibility (so we can read old files, but _not_ backward compatibility -- i.e. allowing _new_ files to be read by _old_ readers -- see below). I would suggest changing the `index_columns` to include dictionaries like

```
{
    'kind': 'range',
    'start': 0,
    'stop': 10,
    'step': 1
}
```

versus

```
{
    'kind': 'serialized',
    'field_name': '__index_level_0__'
}
```

So if a string is encountered in this field (instead of a dict), we know it is "old" metadata. This will break old readers but I think that is OK

Cross ref with https://issues.apache.org/jira/browse/ARROW-1639

cc @cpcloud @martindurant @kszucs @xhochy

### PR DESCRIPTION
**Pull request by wesm**
Closes #25672 

- [x] closes #xxxx
- [x] tests added / passed (N/A)
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff` (N/A)
- [ ] whatsnew entry (N/A)

### CODE REVIEW

File: doc/source/development/developer.rst [hunk 1]
```diff
@@ -53,26 +60,42 @@ for each column, *including the index columns*. This has JSON form:
...
+
+Index Metadata Descriptors
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``RangeIndex`` can be stored as metadata only, not requiring serialization. The
+descriptor format for these as is follows:
+
+.. code-block:: python
+
+   {'kind': range,
```
Comment (TomAugspurger) [inline doc/source/development/developer.rst]: Should the `range` here be s string literal?

Comment (wesm) [reply doc/source/development/developer.rst]: yes, will fix

**Comment by wesm** [comment]
Fixed the range thing

File: doc/source/development/developer.rst [hunk 2]
```diff
@@ -53,26 +60,42 @@ for each column, *including the index columns*. This has JSON form:
...
+See below for the detailed specification for these
+
+Index Metadata Descriptors
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``RangeIndex`` can be stored as metadata only, not requiring serialization. The
+descriptor format for these as is follows:
+
+.. code-block:: python
+
```
Comment (TomAugspurger) [inline doc/source/development/developer.rst]: I guess we lint code-blocks too.

```suggestion

   index = pd.RangeIndex(0, 10, 2)
```

File: doc/source/development/developer.rst [hunk 3]
```diff
@@ -53,26 +60,43 @@ for each column, *including the index columns*. This has JSON form:
...
+   index = pd.RangeIndex(0, 10, 2)
+   {'kind': 'range',
+    'name': index.name,
+    'start': index._start,
```
Comment (jreback) [inline doc/source/development/developer.rst]: https://github.com/pandas-dev/pandas/pull/25720 was just merged, prob could just change this (though of course implementing this for <0.25.x will still require this)

File: doc/source/development/developer.rst [hunk 4]
```diff
@@ -53,26 +60,43 @@ for each column, *including the index columns*. This has JSON form:
...
+    'field_name': '__index_level_0__'}
 
-    .. code-block:: python
+Every index column is stored with a name matching the pattern
+``__index_level_\d+__``. Following this naming convention isn't strictly
+necessary, but strongly suggested for compatibility with Arrow and
+disambiguation. The ``'field_name'`` is the actual name of the column in the
+serialized Parquet table. If the ``Index`` has a non-None ``name`` attribute,
+then it can be found in the ``name`` field of the metadata for that serialized
+data column as described below.
```
Comment (jorisvandenbossche) [inline doc/source/development/developer.rst]: Is this actually correct? (although it was already there in the doc before, to be clear) 
This seems to indicate that an index always gets a ``__index_level_x__`` name as the `field_name`, regardless of the name it has (so not only if it is None).

But this is not what I see from a quick test:

```
In [3]: pyarrow.__version__                                                                                                                                                     
Out[3]: '0.12.0'

In [4]: df = pd.DataFrame({'a': [1, 2, 3], 'b': [3, 4, 5]}).set_index('a')                                                                                                      

In [5]: df                                                                                                                                                                      
Out[5]: 
   b
a   
1  3
2  4
3  5

In [6]: pyarrow.Table.from_pandas(df)                                                                                                                                           
Out[6]: 
pyarrow.Table
b: int64
a: int64
metadata
--------
OrderedDict([(b'pandas',
              b'{"index_columns": ["a"], "column_indexes": [{"name": null, "'
              b'field_name": null, "pandas_type": "unicode", "numpy_type": "'
              b'object", "metadata": {"encoding": "UTF-8"}}], "columns": [{"'
              b'name": "b", "field_name": "b", "pandas_type": "int64", "nump'
              b'y_type": "int64", "metadata": null}, {"name": "a", "field_na'
              b'me": "a", "pandas_type": "int64", "numpy_type": "int64", "me'
              b'tadata": null}], "pandas_version": "0.23.4"}')])
```

(I remember that we had a discussion about this before, but can't directly remember the outcome of that)

Comment (wesm) [reply doc/source/development/developer.rst]: It appears that the current behavior is to use the name of the index if it does not conflict with any of the data columns. So we should update these docs to reflect that

**Comment by wesm** [comment]
To enhance backwards compatibility I'm going to change the `{'kind': 'serialized', 'field_name': $COLUMN}` to simply `$COLUMN`. So only "special" indexes will have the dict representation. Stay tuned

**Comment by jorisvandenbossche** [comment]
I think this is ready to be merged? (at least, I think it reflects the situation in the last pyarrow release).

@martindurant is this OK from the fastparquet side? I think reading such metadata is not yet supported? (https://github.com/dask/fastparquet/issues/414) 
I don't know if fastparquet has plans to also support it when writing parquet files in the long term. If not, we could maybe make this "optional" in the description (as pyarrow can read serialized range indexes fine).

**Comment by martindurant** [comment]
No, this is not yet handled in fastparquet :(
I have found that users of fastparquet are much less likely to ask for a non-useful index to be written, and if there is none when reading, indeed a standard rangeindex is made.

**Comment by jorisvandenbossche** [comment]
> I have found that users of fastparquet are much less likely to ask for a non-useful index to be written

The default is also to not write such indexes in fastparquet, which was probably a good choice. But basically this is a somewhat different way of pyarrow to deal with it, by only storing it in the metadata.

**Comment by jreback** [comment]
@jorisvandenbossche is this ready to merge?

**Comment by jorisvandenbossche** [comment]
Yes, this can be merged now. Did a small clean-up (`_start` -> `start` etc)

**Pull request merged**